src/details/ArborX_DetailsBruteForceImpl.hpp

```diff
@@ -136,8 +142,80 @@ struct BruteForceImpl
               });
         });
   }
+
+  template <class ExecutionSpace, class Predicates, class Values,
+            class Indexables, class Callback>
+  static void query(NearestPredicateTag, ExecutionSpace const &space,
+                    Predicates const &predicates, Values const &values,
+                    Indexables const &indexables, Callback const &callback)
+  {
+    Kokkos::Profiling::ScopedRegion guard("ArborX::BruteForce::query::nearest");
+
+    using MemorySpace = typename Values::memory_space;
+
+    int const n_indexables = values.size();
+    int const n_predicates = predicates.size();
+
+    NearestBufferProvider<MemorySpace> buffer_provider(space, predicates);
+
+    Kokkos::parallel_for(
+        "ArborX::BruteForce::query::nearest::"
+        "check_all_predicates_against_all_indexables",
+        Kokkos::RangePolicy<ExecutionSpace>(space, 0, n_predicates),
+        KOKKOS_LAMBDA(int i) {
+          auto const &predicate = predicates(i);
+          auto const k = getK(predicate);
+          auto const buffer = buffer_provider(i);
+
+          if (k < 1)
+            return;
+
+          using PairIndexDistance =
+              typename NearestBufferProvider<MemorySpace>::PairIndexDistance;
+          struct CompareDistance
+          {
+            KOKKOS_INLINE_FUNCTION bool
+            operator()(PairIndexDistance const &lhs,
+                       PairIndexDistance const &rhs) const
+            {
+              return lhs.second < rhs.second;
+            }
+          };
+
+          PriorityQueue<PairIndexDistance, CompareDistance,
+                        UnmanagedStaticVector<PairIndexDistance>>
+              heap(UnmanagedStaticVector<PairIndexDistance>(buffer.data(),
+                                                            buffer.size()));
+
+          // Nodes with a distance that exceed that radius can safely be
+          // discarded. Initialize the radius to infinity and tighten it once k
+          // neighbors have been found.
+          auto radius = KokkosExt::ArithmeticTraits::infinity<float>::value;
+
+          int j = 0;
+          for (; j < n_indexables && j < k; ++j)
+          {
+            auto const distance = predicate.distance(indexables(j));
+            heap.push(Kokkos::make_pair(j, distance));
+          }
+          for (; j < n_indexables; ++j)
+          {
+            auto const distance = predicate.distance(indexables(j));
+            if (distance < radius)
+            {
+              heap.popPush(Kokkos::make_pair(j, distance));
+              radius = heap.top().second;
+            }
+          }
+          while (!heap.empty())
+          {
+            callback(predicate, values(heap.top().first));
+            heap.pop();
+          }
```

Comment:
We should probably comment that this is sorting the heap.
We technically did not intend to guarantee any order for nearest queries but I suppose we do sort as well in TreeTraversal

Reply:
Except it does not sort the heap in the increasing order. Rather, it is in decreasing order. So, the callbacks here would be called in a different order than in BVH (where they would be called in increasing distance order).

Reply:
Right...
Why did you choose to do that instead of just looping over the elements of the underlying storage?
I know this escapes the control of the data structure but it is more efficient because it skips the heap operations.
(not blocking nor asking you to change at this time, just trying to figure out why you did it this way)

Reply:
Wonder if we should do
```c++
    sortHeap(heap.data(), heap.data() + heap.size(), heap.valueComp());
    for (decltype(heap.size()) i = 0; i < heap.size(); ++i)
      _callback(predicate, values(heap.data() + i)->first);
```
We could skip `sortHeap`, but I wonder if we should. If we don't, we would replicate behavior of the BVH in that the callback will be called in the order from the nearest to further.

Reply:
Implemented with sorting.